streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskSchedulerImpl.scala

```diff
@@ -0,0 +1,180 @@
+
+  /**
+   * This notify the scheduler that {executorId} is failed, and expect a set of
+   * ResourceRequest for all failed tasks on that executor.
+   *
+   * @param executorId executor that failed
+   * @return
+   */
+  def executorFailed(executorId: Int) : Array[ResourceRequest]
+}
+
+class TaskSchedulerImpl(appId : Int, config: Config)  extends TaskScheduler {
+  private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
+
+  private var tasks = Map.empty[TaskId, TaskLaunchData]
+
+  // maintain a list to tasks to be scheduled
+  private var tasksToBeScheduled = Map.empty[Locality, mutable.Queue[TaskLaunchData]]
+
+  // find the locality of the tasks
+  private val taskLocator = new TaskLocator(config)
+
+  private var executorToTaskIds = Map.empty[Int, Set[TaskId]]
+
+  def setTaskDAG(dag: DAG): Unit = {
+    dag.tasks.foreach { params =>
+      val (taskGroupId, taskDescription) = params
+      0.until(taskDescription.parallelism).map((taskIndex: Int) => {
+        val taskId = TaskId(taskGroupId, taskIndex)
+        val locality = taskLocator.locateTask(taskDescription)
+        val taskLaunchData = TaskLaunchData(taskId, taskDescription, dag.subGraph(taskGroupId))
+        tasks += (taskId -> taskLaunchData)
+        scheduleTaskLater(taskLaunchData, locality)
+      })
+    }
+    val nonLocalityQueue = tasksToBeScheduled.getOrElse(NonLocality, mutable.Queue.empty[TaskLaunchData])
+    //reorder the tasks to distributing fairly on workers
+    val taskQueue = nonLocalityQueue.sortBy(_.taskId.index)
+    tasksToBeScheduled += (NonLocality -> taskQueue)
+  }
+
+  def getResourceRequests(): Array[ResourceRequest] ={
+    fetchResourceRequests(fromOneWorker = false)
+  }
+
+  import Relaxation._
+  private def fetchResourceRequests(fromOneWorker: Boolean = false): Array[ResourceRequest] ={
+    var resourceRequests = Array.empty[ResourceRequest]
+    tasksToBeScheduled.foreach(taskWithLocality => {
+      val (locality, tasks) = taskWithLocality
+      LOG.debug(s"locality : $locality, task queue size : ${tasks.size}")
+
+      if (tasks.size > 0) {
+      locality match {
+        case locality: WorkerLocality =>
+          resourceRequests = resourceRequests :+
+            ResourceRequest(Resource(tasks.size), locality.workerId, relaxation = SPECIFICWORKER)
+        case _ if fromOneWorker =>
+            resourceRequests = resourceRequests :+ ResourceRequest(Resource(tasks.size), relaxation = ONEWORKER)
+        case _ =>
+            resourceRequests = resourceRequests :+ ResourceRequest(Resource(tasks.size))
+        }
+      }
+    })
+    resourceRequests
+  }
+
+  def resourceAllocated(workerId : Int, executorId: Int) : Option[TaskLaunchData] = {
+    val locality = WorkerLocality(workerId)
+
+    // first try to schedule with worker locality. If not found, search tasks without locality
+    val task = scheduleTaskWithLocality(locality).orElse(scheduleTaskWithLocality(NonLocality))
+    task match {
```

Comment:
But it may not apply everywhere. 

If it is one sentence operation, it is more reable, like:
option.map(dosometing(_))

If the opeation is complex, I will prefer:
option match {
 case Some(value) =>
   //a long and complex operation.

 case None=>
}

Readbility is more important than shorter code.But it may not apply everywhere. 

If it is one sentence operation, it is more reable, like:
option.map(dosometing(_))

If the opeation is complex, I will prefer:
option match {
 case Some(value) =>
   //a long and complex operation.

 case None=>
}

Readbility is more important than shorter code.